src/opensynth/models/faraday/new_gmm/new_gmm_model.py

```diff
@@ -281,73 +284,104 @@ def training_step(self, batch) -> None:
             precision_cholesky=precision_cholesky,
-            covariances=covar,
-            log_prob=log_prob,
+            covariances=covariances,
+            nll=torch.neg(log_prob),
         )
 
+        self.log(
+            "nll",
+            torch.neg(log_prob),
+            on_step=False,
+            on_epoch=True,
+        )  # uses mean-reduction (default) to accumulate the metrics
+
+    def on_train_batch_end(self, *args, **kwargs) -> None:
+        if self.sync_on_batch:
+            weights = self.gmm_module.weights
+            means = self.gmm_module.means
+            precision_cholesky = self.gmm_module.precision_cholesky
+            covariances = self.gmm_module.covariances
+            nll = self.gmm_module.nll
+
+            # forward performs update, compute and reset metrics
+            weights_reduced = self.weight_metric.forward(weights)
+            means_reduced = self.mean_metric.forward(means)
+            prec_chol_reduced = self.precision_cholesky_metric.forward(
+                precision_cholesky
+            )
+            covar_reduced = self.covariance_metric.forward(covariances)
+            nll_reduced = self.nll.forward(nll)
+
+            print("Batch NLL: ", nll_reduced)
+
+            self.gmm_module.update_params(
+                weights=weights_reduced,
+                means=means_reduced,
+                precision_cholesky=prec_chol_reduced,
+                covariances=covar_reduced,
+                nll=nll_reduced,
+            )
+
     def on_train_epoch_end(self) -> None:
         # At the end of epoch, update metrics and sync across
         # multiple devices using torchmetrics.Metric.compute
         # Then update model params using the synced values
 
-        weights = self.gmm_module.weights
-        means = self.gmm_module.means
-        precision_cholesky = self.gmm_module.precision_cholesky
-        covariances = self.gmm_module.covariances
-        log_prob = self.gmm_module.log_prob
+        if not self.sync_on_batch:
+            weights = self.gmm_module.weights
+            means = self.gmm_module.means
+            precision_cholesky = self.gmm_module.precision_cholesky
+            covariances = self.gmm_module.covariances
+            nll = self.gmm_module.nll
 
-        print(
-            f"Local weights at rank: {self.local_rank} -",
-            f"means: {weights[0]:.4f}, {means[0][0]:.4f}",
-        )
-        self.weight_metric.update(weights)
-        self.mean_metric.update(means)
-        self.precision_cholesky_metric.update(precision_cholesky)
-        self.covariance_metric.update(covariances)
-        self.log_prob.update(log_prob)
+            self.weight_metric.update(weights)
+            self.mean_metric.update(means)
+            self.precision_cholesky_metric.update(precision_cholesky)
+            self.covariance_metric.update(covariances)
+            self.nll.update(nll)
 
         weights_reduced = self.weight_metric.compute()
```

Comment:
Should all this code also be indented inside the if statement?

Reply:
Yes, although this won't make a different to the code output, it will just remove a repeated compute() as this already happened on batch end, when `sync_on_batch=True`.